### ISSUE TITLE
How exactly should "parametric nullness" be defined? [working decision: no change]

### ISSUE THREAD

**Issue by kevinb9n** [jspecify/jspecify#275]
Using the example `@NullMarked class Foo<A extends @Nullable Object, B>`...

We have defined "parametric nullness" as a special 4th nullness indicator, which `A` has here, having the unique nature "null can come out but can't go in".

```
A a;
methodAcceptingNonNull(a); // problem
a = null; // also a problem
```

(A variation, not being debated here, is that `B` also has parametric nullness, and now its nature is just that "null can *maybe* come out but can't go in; check the bound to see.")

**Note:** These arguments below depend on a positive response to #274 (so consider skimming that first, and if the tide goes against it, disregard this too).

That issue (post plus first reply) gives a sufficient explanation for why you can't assign `null` to such a type, without any need to appeal to a special "kind" of nullness. That rule will probably be internalized as "you just can't assign a `null` literal to a type-variable type, sorry" (but at least it does flow naturally from basic principles and isn't as arbitrary as that sounds).

But with that rule established _generally_, then do the ways we describe `A` and `B` really need to keep accounting for it over and over? It would seem fair to describe `A` as nullable and `B` as non-null, plain and simple, and the general rule takes care of the rest.

Of course, I see a lot of benefit in that, being able to reduce these to the same categories we already have. Some costs though:

* We don't get to say nice tidy things like "a non-null type is assignable to the nullable form of that type". The general rule about type variables has to be checked too.
* It seems terminally confusing to call `A` nullable yet `@Nullable A` still means something distinct! We would try to explain that it really means a union type, _definitely_ adding `null` to the value set whereas it was only _possibly_ there otherwise, but there's only so much we could do.

---

If we were to make that shift, then there is still some use for defining "parametric nullness" too, but with a more basic meaning. It would now mean "it does have a definite nullness... you'd just have to go look up the type parameter bound to know which it is." (This may be what you already figured it meant...)

So calling `A` "nullable" or saying it's "of parametric nullness" might _both_ be correct (and analogously for `B`). It just depends whether you want to include the non-local information or not.

(We're still left us with notational problems, just maybe slightly different from the ones we currently have. That's #129.)

**Comment by netdpb** [comment]
> That issue (post plus first reply) gives a sufficient explanation for why you can't assign `null` to such a type

I think #274 does lead to that conclusion, but it doesn't contain a pithy explanation. You state there that "`String` means actual string", and then talk about `@Nullable ___` and `@NonNull ___` working the same whether `___` is a type variable or not. And in the second comment, you state "`T` does _not_ include null` but later suggest that that only refers to the fact that you cannot assign literal `null` to them.

How would you phrase the general rule about type variables?

I still feel like my mental model treats type variables in relation to non-variable types as variables are to literals. `int x` might be assigned with 0, so don't divide by it without checking. `T` might be substituted with a nullable type, so don't do `Optional<T>`, do `Optional<@NonNull T>`; alternatively, it might be substituted with a non-null type, so don't dereference variables of type `T` without checking for null and don't assign `null` to any such variables. (Just like I can't return `"foo"` from a method whose return type is `T` where `T extends CharSequence`.)

To me, that mental model allows nullability to be a straightforward application of how I think about type variables to the concept of nullability of a type. And it doesn't really require a new kind of nullability, which I agree is nice.

**Comment by kevinb9n** [comment]
I'm trying to answer the parts that are about #274 on #274.


> I still feel like my mental model treats type variables in relation to non-variable types as variables are to literals. `int x` might be assigned with 0, so don't divide by it without checking. `T` might be substituted with a nullable type, so don't do `Optional<T>`, do `Optional<@NonNull T>`; alternatively, it might be substituted with a non-null type, so don't dereference variables of type `T` without checking for null and don't assign `null` to any such variables. (Just like I can't return `"foo"` from a method whose return type is `T` where `T extends CharSequence`.)

Strangely, this is also my answer for "how do I think about a type variable". I've been saying: it's a local type whose goal is to preserve its substitutability for any type in its bounds, so it prevents you from doing *whatever* it has to to achieve that.

So if you're stating this in a spirit of contradiction to something I've said, we might be miscommunicating.

It is *basic*, though, and still leaves you having to think through a lot of stuff the long way. The less frequently you have to actually sit and try to range your mind over all these possible cases, the better, I think. For example, the first comment on #274 suggests two valid explanations for a question, with one of them based pretty directly on your model. But it's not "pithy"; the other one is more so.

**Comment by netdpb** [comment]
In the first comment, you said:

> (A variation, not being debated here, is that B also has parametric nullness, and now its nature is just that "null can maybe come out but can't go in; check the bound to see.")

I think that variation (which I was afraid to debate here) leads to a simpler way to think about your conundrum.

* `A`, being an unadorned type variable usage just like `B`, has parametric (I would prefer to call it "variable") nullness, like all unadorned type variables. It is not inherently nullable; it's just that its bound doesn't stop truly nullable types from substituting for it.
* `@Nullable A` _is_ nullable because it's a nullable projection of a type variable.

`A` and `@Nullable A` have distinct nullness (even though they evaluate to the same thing). And contraily, `A` and `B` have the _same_ nullness. You don't have to explain why references that look different are actually the same or why references that look the same are actually different.

**Comment by kevinb9n** [comment]
> `A` and `B` have the _same_ nullness.

Yet one can be passed to `foo(@NonNull Object)` and the other can't. Which suggests that instead of saying just

"You can only pass a non-null expression to a non-null parameter"

we would have to say 

"You can only pass either a non-null expression, or an expression whose type is a type variable that has a non-null upper bound, to a non-null parameter".

As for why I think this is fatal, I'd better unpack:

### Why talk about "a nullness" at all?

My goal has been a definition of "the nullness of a type usage" that:

* is complete (every type usage has one)
* is enumerated (there's only a short list of different ones to be had)
* is useful (we should be able to understand why a finding is or isn't generated while appealing only to the nullness of each type usage involved)

I think both of our proposals seem to fail this test. But the current definition (as written in the glossary) seems to pass it.

### Proposed resolution

Close with no change; keep it the way it's been. In null-marked code:

`@NullMarked class Foo<A extends @Nullable Object, B>`

| type usage of | if explicit `@Nullable` | if explicit `@NonNull` | if no annotation |
| :-: | :-: | :-: | :-: |
| `Foo` | nullable | non-null | non-null |
| `B` | nullable | non-null | non-null |
| `A` | nullable | non-null | parametric |

This table illustrates a few things:

* The explicit annotation is always respected
* `String` and `B` will be treated identically, nullness-wise
* `@NonNull String` and `@NonNull B` are the two (and only two) redundancies

### Parametric vs. variable

I don't mind "variable nullness" itself, but I worry it would be confused for "the nullness of a variable". It also suggests that `A` *itself* might sometimes be nullable and sometimes not, when really its nullness is always the same.

**Comment by kevinb9n** [comment]
The rule for automated type conversion between these three nullnesses is, I guess:

"Any nullness is convertible to nullable, and non-null is convertible to any nullness, and parametric is convertible to parametric."

I think(?) that last part works, but only because we can assume that base type conversion already succeeded. So it must be a case like `A` to `A`, or `D` to `C` where

`class Bar<C extends @Nullable Object, D extends C>`

Notice I'm leaving unspec out of this whole thing, because I still think its behavior should be governed by "many worlds" as often as possible.

**Comment by cpovirk** [comment]
> Parametric is convertible to parametric

Sadly, there is also the case:

`class Bar<C extends @Nullable Object, D extends @Nullable C>`

But even if the rules above aren't "useful" in the sense of providing the answer to every possible question, I think they're "useful" in the sense of being, well, useful :) They do (I think) give the right answer for cases other than parametric vs. parametric. And I like the idea that, of anything is special enough to deserve its own term, it's the `T extends @Nullable Object` case.

**Comment by kevinb9n** [comment]
Ah hell.

**Comment by netdpb** [comment]
> > `A` and `B` have the _same_ nullness.
> 
> Yet one can be passed to `foo(@NonNull Object)` and the other can't.

Isn't this exactly like the fact that of `<A extends Object, B extends CharSequence>` only one can be passed to `foo(CharSequence)`? And yet we don't say "the type of `B` is `CharSequence`". I think our explanation can rely on this well-known property of type variables.

> Which suggests that instead of saying just
> 
> "You can only pass a non-null expression to a non-null parameter"
> 
> we would have to say
> 
> "You can only pass either a non-null expression, or an expression whose type is a type variable that has a non-null upper bound, to a non-null parameter".

Hmm. When we say "You can only pass a `CharSequence` (expression) to a `CharSequence` parameter", we do not mean to exclude `String` expressions. So "You can only pass a non-null expression to a non-null parameter" can include expressions whose static type is _known to be convertible to non-null_, which includes `<A extends Object>`, without calling `A` non-null.

Maybe I'm distinguishing without difference, but when I think of parametric/variable nullness I am not thinking really of a distinct kind of nullness but more of the natural essence of any type variable: the actual type that will be substituted is constrained by its upper bound, and developers _always_ have to think about that when interacting with the type variable.

* `T extends CharSequence` _is not_ the same as `CharSequence`, but all substituted types must be subtypes of `CharSequence`.
   * `<T extends @Nullable CharSequence>` _is not_ nullable; both nullable and non-nullable subtypes of `CharSequence` can be substituted for it.
   * `<T extends CharSequence>` _is not_ non-nullable, but all substituted types must be.
* `T extends CharSequence` might be substituted with `CharSequence` or a convertible type such as `String`; using code must handle any such type.
   * `<T extends @Nullable CharSequence>` might be substituted with `@Nullable CharSequence` or a convertible type such as `CharSequence` or `@Nullable String`; using code must handle any such type, nullable or not.

-----

> > Parametric is convertible to parametric
> 
> Sadly, there is also the case:
> 
> `class Bar<C extends @Nullable Object, D extends @Nullable C>`

The story for convertibility between `C` and `D` is that `D` can't be converted to `C` without a nullness check, since `D` might be substituted with a nullable type and `C` with a non-null type. Right? And you have to look at the upper bounds to figure that out.

**Comment by kevinb9n** [comment]
> > > `A` and `B` have the _same_ nullness.
> > Yet one can be passed to `foo(@NonNull Object)` and the other can't.
> 
> Isn't this exactly like the fact that of `<A extends Object, B extends CharSequence>` only one can be passed to `foo(CharSequence)`? And yet we don't say "the type of `B` is `CharSequence`". I think our explanation can rely on this well-known property of type variables.

I think there is an important difference, which explains what led me to the three goals under "why talk about nullness at all" above in the first place.

Among base types there is a large and open-ended type hierarchy. To know whether two types are assignable, you just have to consult that hierarchy, pretty much every time. "Check whether the from-type is a subtype of (yeah, or otherwise assignable to) the to-type." There's no shortcut.

To preserve the analogy you want would suggest speaking to users like this: "Find the nullness component of the from-type and of the to-type, then check the hierarchy and see whether the first is a subtype of the second."

I don't want to talk like that. The difference is that, here, that hierarchy is small and fixed. That is a different ball game. We have the ability to speak much less abstractly and more understandably, and that's what produced those aforementioned three goals.


> Hmm. When we say "You can only pass a `CharSequence` (expression) to a `CharSequence` parameter", we do not mean to exclude `String` expressions. So "You can only pass a non-null expression to a non-null parameter" can include expressions whose static type is _known to be convertible to non-null_, which includes `<A extends Object>`, without calling `A` non-null.

I just think this is unnecessarily convoluted.


> Maybe I'm distinguishing without difference _[To be very clear, if my perspective has stopped being useful because it's not resonating with anyone, please tell me and I will stop; my comments are not meant to block progress.]_, but when I think of parametric/variable nullness I am not thinking really of a distinct kind of nullness but more of the natural essence of any type variable: the actual type that will be substituted is constrained by its upper bound, and developers _always_ have to think about that when interacting with the type variable.

Yeah, this is reminiscent of the top comment where I said "But with that rule established *generally*, then do the ways we describe `A` and `B` really need to keep accounting for it over and over?" I've come around to "yeah, they should" and spelled out my reasons.

I had guessed that this part would be persuasive to you:

> Another body blow to including `B` as "parametric" is that `@NonNull B` is classified as a redundancy (the annotation's presence or absence has no consequence on the information content being captured), so it would be weird and problematic if the nullness of the type hinged on it.

but you didn't respond to it. imho, it can't make sense for redundant information to change the nullness, and it can't make sense to say that `@NonNull B` is somehow not redundant.

And you also haven't really said what you feel is the _problem_ with the proposed resolution. I described its nice properties; what are its unnice properties, or why are those properties not nice?

Also I think the proposed resolution is the only thing I've seen that makes #129 tidy.

**Comment by kevinb9n** [comment]
To go back several comments, where I said "whether parametric is convertible to parametric is complicated"...

Leaving it like that is unsatisfying. So maybe the way to unpack it is:

"When we see `T%` and `U%` we just have to bear in mind that each is parametric along its own generally-independent axis. That is, parameterizations could include any possible combinations of type-argument nullnesses. Or, when there are dependencies between them -- of the form "`T` is no more nullable than `U` is" -- those just explain why *more* conversions worked than you thought would, not fewer, so it shouldn't be a big worry. And at least any `%` will relate to each of `?/!/*` in the same consistent way as any other `%`."

Is that right?

If a person is at the point of wanting to understand why they couldn't assign a `T%` to a `U%`, then perhaps they're ready to read and understand that?

**Comment by cpovirk** [comment]
I think that's all fine with me:
- I'm happy with the definition that produces the table in https://github.com/jspecify/jspecify/issues/275#issuecomment-1244636794.
- I'm happy with the term "parametric nullness."
- Parametric nullness interacts in simple ways with non-parametric nullness.
- "Parametric-parametric conversions mostly aren't allowed (in the same way that plain-Java assignment between `T` and `U` typically isn't allowed), but there are exceptions" sounds like a model that will work well enough in practice for users.
- The existence of exceptions means that you can't answer all questions just from the nullnesses of the types involved (because you may still need to consult bounds). But that's OK, and "parametric nullness" remains a useful addition because (among other things) it _often_ answers users' questions.

If there are any loose ends here, they might be around _partially_ annotated code? But that's probably best left for #248 or elsewhere, since there's enough going on here as it is. (Maybe you and netdpb still have something hanging?)

**Comment by netdpb** [comment]
> I had guessed that this part would be persuasive to you:
> 
> > Another body blow to including `B` as "parametric" is that `@NonNull B` is classified as a redundancy (the annotation's presence or absence has no consequence on the information content being captured), so it would be weird and problematic if the nullness of the type hinged on it.
> 
> but you didn't respond to it. imho, it can't make sense for redundant information to change the nullness, and it can't make sense to say that `@NonNull B` is somehow not redundant.

I think that there's a useful difference between nullness information available directly from the type variable usage itself and nullness information derived from the type variable's definition elsewhere, which is why `@NonNull B` doesn't strike me as redundant in the way that `@NonNull int` does. (I recognize that information from the type variable definition is still in the same compilation unit and therefore closer than information about the upper bound type.)

> Among base types there is a large and open-ended type hierarchy. To know whether two types are assignable, you just have to consult that hierarchy, pretty much every time. "Check whether the from-type is a subtype of (yeah, or otherwise assignable to) the to-type." There's no shortcut.
> 
> To preserve the analogy you want would suggest speaking to users like this: "Find the nullness component of the from-type and of the to-type, then check the hierarchy and see whether the first is a subtype of the second."
> 
> I don't want to talk like that. The difference is that, here, that hierarchy is small and fixed. That is a different ball game. We have the ability to speak much less abstractly and more understandably, and that's what produced those aforementioned three goals.

I think your point about nullness being a closed system rather than an open one like subtypes is important. Elsewhere I've said that the question at hand is one of _convertibility_ and not general _subtyping_ and that it might be counterproductive to refer to `@NonNull String` as a _subtype_ of `@Nullable String`, much as `int` is convertible to `Integer` without being a subtype of it. I don't think it's necessary to think about type variable substitutions as being just any subtype of the upper bound, where the nullness component just gets subsumed into the general idea of subtypes. It may be possible to describe the variable nullness of type variables more specifically than that while still referring to the upper bound.

That being said, while I suspect that I might be able to write up an alternative way of thinking about parametric nullness, I'm not claiming that the way you're proposing to describe it is bad or worse than we already have. If I get around to it, I'd propose writing up this alternative framing as an optional description for people for whom it would be helpful, but I don't think it's blocking anything.

Go forth and parameterize.

**Comment by kevinb9n** [comment]
I'm going to concede a flaw, but still propose to resolve this as planned.

> I think that there's a useful difference between nullness information available directly from the type variable usage itself and nullness information derived from the type variable's definition elsewhere, which is why `@NonNull B` doesn't strike me as redundant in the way that `@NonNull int` does. (I recognize that information from the type variable definition is still in the same compilation unit and therefore closer than information about the upper bound type.)

"Uh oh"

Indeed my reasoning has at times leaned on the idea that this is all local to one compilation unit. But, well ...

```java
// CompilationUnit1.java
class Foo<T> {}

// CompilationUnit2.java
class Bar<S> extends Foo<T> {}
```

So my proposed resolution is in fact conceding that the nullness of `S` is only as deterministic as this file's dependencies are. If we know the deps, we know the nullness; if we don't, we don't.

I'm still going to argue that this is appropriate. I would say that any conceptualization of "the nullness of `S`" that *isn't* dependent on CompilationUnit1... would also be pretty meaningless. It wouldn't tell us anything very useful about `S`. I would say that the meaningful nullness properties of `S` simply *do* depend on the other file, and our terminology shouldn't pretend otherwise.

**Comment by kevinb9n** [comment]
Botched the code, rethinking.

**Comment by kevinb9n** [comment]
Generics are hard.

I think I'm talking crap. The declaration of `<S>` -- even if it extends some other type variable, even if it doesn't -- definitively decides its own nullness.  It never just depends on the other file. Does it?

**Comment by kevinb9n** [comment]
Suggest rewording:

Type variables whose upper bound is non-null are considered non-null; all the rest have only "parametric nullness".

**Comment by netdpb** [comment]
@kevinb9n Is that rewording about changing the order of the cases, or about using the term "type variable" rather than "type parameter"?

What about this? "Unprojected type variable usages in null-marked scope whose upper bound is non-nullable are considered non-nullable. Unprojected type variable usages in null-marked scope whose upper bound is nullable have parametric nullness."

Projected type variable usages never have parametric nullness; unspecified ones are unspecified; unprojected ones  with unspecified bounds are unspecified, so it's not true that "the rest" have parametric nullness.

**Comment by kevinb9n** [comment]
Ah, my mistake. Your latest statement is good but is only addressing 2 of the 4 cases for the type variable's bound.

"In null-marked context, what is the effective nullness of an unprojected usage of a type variable?

If the type variable's bound is:
* unspecified --> it is unspecified
* non-null --> it is non-null
* nullable --> it is parametric
* parametric --> it is parametric"

**Comment by cpovirk** [comment]
The unspecified part takes us back to the discussion going in https://github.com/jspecify/jspecify/issues/238#issuecomment-2166589665. Aside from that, I suspect that there's agreement.

**Comment by netdpb** [comment]
This issue is only about parametric nullness; we can discuss unspecified elsewhere. I've made an edit to the [proposal](https://github.com/jspecify/jspecify/issues/275#issuecomment-2045981298). Please take another look.
